pkg/image/docker/manifest.go

```diff
@@ -1,38 +1,147 @@
 	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/google/go-containerregistry/pkg/v1/types"
 )
 
-type manifest struct {
-	raw    []byte
+type dockerManifest struct {
 	parsed tarball.Manifest
 }
 
 // newManifest creates a new manifest struct from the given Docker archive manifest bytes
-func newManifest(raw []byte) (manifest, error) {
+func newManifest(raw []byte) (*dockerManifest, error) {
 	var parsed tarball.Manifest
 	if err := json.Unmarshal(raw, &parsed); err != nil {
-		return manifest{}, fmt.Errorf("unable to parse manifest.json: %w", err)
+		return nil, fmt.Errorf("unable to parse manifest.json: %w", err)
 	}
 
 	if len(parsed) == 0 {
-		return manifest{}, fmt.Errorf("no valid manifest.json found")
+		return nil, fmt.Errorf("no valid manifest.json found")
 	}
 
-	return manifest{
-		raw:    raw,
+	return &dockerManifest{
 		parsed: parsed,
 	}, nil
 }
 
-// tags returns the image tags referenced within the images manifest file (within the given docker image tar).
-func (m manifest) tags() (tags []string) {
+// allTags returns the image tags referenced within the images manifest file (within the given docker image tar).
+func (m dockerManifest) allTags() (tags []string) {
 	for _, entry := range m.parsed {
 		tags = append(tags, entry.RepoTags...)
 	}
 	return tags
 }
+
+// extractManifest is helper function for extracting and parsing a docker image manifest (V2) from a docker image tar.
+func extractManifest(tarPath string) (*dockerManifest, error) {
+	f, err := os.Open(tarPath)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err := f.Close()
+		if err != nil {
+			log.Errorf("unable to close tar file (%s): %w", f.Name(), err)
+		}
+	}()
+
+	manifestReader, err := file.ReaderFromTar(f, "manifest.json")
+	if err != nil {
+		return nil, err
+	}
+
+	contents, err := ioutil.ReadAll(manifestReader)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read manifest.json: %w", err)
+	}
+	return newManifest(contents)
+}
+
+// generateOCIManifest takes a docker manifest and a path to the tar and generates an OCI manifest derived from the given arguments and the docker config.
+func generateOCIManifest(tarPath string, manifest *dockerManifest) (*v1.Manifest, error) {
+	f, err := os.Open(tarPath)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		err := f.Close()
+		if err != nil {
+			log.Errorf("unable to close tar file (%s): %w", f.Name(), err)
+		}
+	}()
+
+	if len(manifest.parsed) != 1 {
+		return nil, ErrMultipleManifests
+	}
+
+	configReader, err := file.ReaderFromTar(f, manifest.parsed[0].Config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to find docker config: %w", err)
+	}
+
+	configContents, err := ioutil.ReadAll(configReader)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read docker config: %w", err)
+	}
+
+	var layerSizes = make([]int64, len(manifest.parsed[0].Layers))
+	for idx, layerTarPath := range manifest.parsed[0].Layers {
```

Comment:
Another comment in the category of "doesn't mean this should change, but it reminded me of a pattern I've noticed recently":

For these "map" style implementations, where we're converting each item in a collection to another kind of item in a result collection, I've been seeing people `make` the slice by specifying the slice capacity using `len(...)` instead of the length (and the length is specified as `0`). The effect this has is the `for` loop ends up looking more like a "normal" loop where the index isn't needed, and you can just `append` as necessary. This is equally as efficient in memory allocation as the index-oriented approach, since the entire block of underlying memory is still allocated up front, it's just the _slice_ that's "growing" (without causing a reallocation of the underlying _array_) upon each iteration.

Not sure if I like this approach better or not, I just found it interesting.